### DISCUSSION TITLE
Blank, gray screen from web node when using Firefox (concourse-7.0.0)

### DISCUSSION THREAD

**Discussion by cwses1** [concourse/concourse#6645] [Help & Support] [answered]
Dear Concourse Community:

My Dashboard (http://127.0.0.1:8080/) is coming up as a blank screen with a dark gray background.  I was expecting much more.  Please see the following, almost totally useless screenshot:

![Concourse_BlankGrayPageScreenshot](https://user-images.githubusercontent.com/14972218/110207398-44ffd600-7e51-11eb-9f96-012d75ed76f1.png)

You will notice that the 'concourse web' server is running well because the 'Concourse' title and icon are coming through - and I am also able to view the source HTML.  Before writing anything else, I'll share the details of my install:

concourse-7.0.0-windows-amd64
fly-7.0.0-windows-amd64
Firefox 86.0
PostgreSQL 13.2

This appears to be a client-side issue.  Looking at the browser's console errors I see this:

`The resource from “http://127.0.0.1:8080/public/clipboard.min.js?id=3f3688138a1b9fc4ef669ce9056b6674” was blocked due to MIME type (“text/plain”) mismatch (X-Content-Type-Options: nosniff).
127.0.0.1:8080`

`Loading module from “http://127.0.0.1:8080/public/elm-setup.js?id=f735373f5756c8e9939ad27e74a7b96b” was blocked because of a disallowed MIME type (“text/plain”).
127.0.0.1:8080`

`The resource from “http://127.0.0.1:8080/public/elm.min.js?id=a998f43ebe9394aaf25b0d5c426b9d07” was blocked due to MIME type (“text/plain”) mismatch (X-Content-Type-Options: nosniff).
127.0.0.1:8080`

`The resource from “http://127.0.0.1:8080/public/clipboard.min.js?id=3f3688138a1b9fc4ef669ce9056b6674” was blocked due to MIME type (“text/plain”) mismatch (X-Content-Type-Options: nosniff).
127.0.0.1:8080`

Firefox is complaining only about these JavaScript files.  The content types of these of the JS files are set to text/plain and have a size of 0 B.  See the Type Column below:

![Concourse_ContentTypes](https://user-images.githubusercontent.com/14972218/110207599-978dc200-7e52-11eb-8ba7-330e4915cb08.png)

Let's examine the headers of the first one (elm.min.js):

![Concourse_ResponseHeaders](https://user-images.githubusercontent.com/14972218/110207668-0e2abf80-7e53-11eb-9c11-7237719b2a61.png)

The error messages from above points us to two important headers:

`Content-Type: text/plain; charset=utf-8`
`X-Content-Type-Options: nosniff`

The web server is telling the browser to block the script if it doesn't have a JavaScript MIME Type.  For additional details, please see this:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options

Therefore, Firefox is refusing to process the script files.

I feel like this issue could be resolved if Concourse returned the correct content type for the script files - or if I could configure Firefox to ignore the content type options - but I cannot figure out how to do either option.

This issue also applies to Edge, which gives us a more descriptive error message:

`(index):1 Refused to execute script from 'http://127.0.0.1:8080/public/elm.min.js?id=a998f43ebe9394aaf25b0d5c426b9d07' because its MIME type ('text/plain') is not executable, and strict MIME type checking is enabled.`

I did not test this against Chrome.

**Comment by taylorsilva** [comment]
This is very odd. I use firefox as well, version 86. Guess my difference is that I'm on macOS. I brought up Concourse locally using the [quickstart docker-compose](https://concourse-ci.org/quick-start.html). I get no errors in my browsers console and the page loads as expected.
<img width="1274" alt="image" src="https://user-images.githubusercontent.com/4583837/110376272-fe60d600-8020-11eb-9135-17661efea3f1.png">


If you visit https://ci.concourse-ci.org/ do you also get a blank screen? How did you deploy your Concourse web node; locally on your windows machine?

**Reply by cwses1** [reply]
No, I do not get a blank screen for https://ci.concourse-ci.org/.  It works beautifully just like every other installation I've ever worked with!  See here:

![Concourse_Working](https://user-images.githubusercontent.com/14972218/110468090-cfd90e80-80a5-11eb-9c08-c26c2bd0c23b.png)

I believe this is because the server is reporting the correct JavaScript content types.  Notice the difference:

![Concourse_CorrectContentTypes](https://user-images.githubusercontent.com/14972218/110468323-1fb7d580-80a6-11eb-94d8-97bd750fa8c0.png)

No console errors either in the browser.

I set up Concourse manually - not using Docker Compose.  Postgres runs on Windows natively.  The Concourse web and worker are running in the Windows console and configured using environment variables.

**Comment by chenbh** [comment, marked as answer]
Ohh boy, Windows. So golang tries to figure out the `content-type` based on the file names, and it looks like Windows managed to botch that up for `.js` files. A quick skim of [this stackoverflow answer](https://stackoverflow.com/a/65150421) and the corresponding bug report on [golang itself](https://github.com/golang/go/issues/32350) sounds like this is really out of our (and even golang's) control.

My suggestion would be trying out the registry edit solution, or running the web process from a Mac or Linux machine

**Reply by cwses1** [reply]
Thanks @chenbh !  This is the solution for Windows.  Let's look at the registry at `Computer\HKEY_CLASSES_ROOT\.js`:

![Concourse_RegistryWrong](https://user-images.githubusercontent.com/14972218/110472125-f9486900-80aa-11eb-9f59-fc185379a89a.png)

You will see that `.js` has a `Content Type` of `text/plain`.  That doesn't make any sense.  Let's fix that

![Concourse_RegistryWrongButFixed](https://user-images.githubusercontent.com/14972218/110472591-8be90800-80ab-11eb-8d55-562be2123428.png)

and restart our `concourse web` command.  Results:

![Concourse_WorkingLocally](https://user-images.githubusercontent.com/14972218/110472762-c3f04b00-80ab-11eb-9b08-b6d4f85c1c2c.png)

No blank, gray screen!  It works!  So this was a success.

And when I change it back to `text/plain`, restart `concourse web`, and reload, it goes back to the blank, gray screen.

**OTHER DETAILS**

You must restart `concourse web` after you change the registry to see the effects.  That is, the effects are not immediate when altering the registry.

As written in the Stack Overflow article you referenced, there are two locations where you can define content types in the registry.  The first location had the wrong content type.  The second location `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.js' has the correct content type and does not have to be modified.

![Concourse_RegistryCorrect](https://user-images.githubusercontent.com/14972218/110471856-a66eb180-80aa-11eb-84d0-9be2c77a4111.png)

So the new rule is, "If you get a blank, gray screen on Windows - fix the registry."
